bugsnag-plugin-android-ndk/src/main/java/com/bugsnag/android/ndk/NativeBridge.kt

```diff
@@ -0,0 +1,164 @@
+package com.bugsnag.android.ndk
+
+import android.os.Build
+import android.util.Log
+import com.bugsnag.android.NativeInterface
+import com.bugsnag.android.StateEvent
+import com.bugsnag.android.StateEvent.*
+import java.io.File
+import java.util.Observable
+import java.util.Observer
+import java.util.UUID
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.locks.ReentrantLock
+
+/**
+ * Observes changes in the Bugsnag environment, propagating them to the native layer
+ */
+class NativeBridge : Observer {
+
+    private val lock = ReentrantLock()
+    private val installed = AtomicBoolean(false)
+    private val reportDirectory: String = NativeInterface.getNativeReportPath()
+
+    private val is32bit: Boolean
+        get() {
+            val abis = NativeInterface.getCpuAbi()
+            return !abis.toList().any { it.contains("64") }
+        }
+
+    external fun install(
+        reportingDirectory: String,
+        autoDetectNdkCrashes: Boolean,
+        apiLevel: Int, is32bit: Boolean,
+        appVersion: String,
+        buildUuid: String,
+        releaseStage: String
+    )
+
+    external fun startedSession(
+        sessionID: String,
+        key: String,
+        handledCount: Int,
+        unhandledCount: Int
+    )
+
+    external fun deliverReportAtPath(filePath: String)
+    external fun addBreadcrumb(name: String, type: String, timestamp: String, metadata: Any)
+    external fun addMetadataString(tab: String, key: String, value: String)
+    external fun addMetadataDouble(tab: String, key: String, value: Double)
+    external fun addMetadataBoolean(tab: String, key: String, value: Boolean)
+    external fun addHandledEvent()
+    external fun addUnhandledEvent()
+    external fun clearBreadcrumbs()
+    external fun clearMetadataTab(tab: String)
+    external fun removeMetadata(tab: String, key: String)
+    external fun pausedSession()
+    external fun updateContext(context: String)
+    external fun updateInForeground(inForeground: Boolean, activityName: String)
+    external fun updateOrientation(orientation: Int)
+    external fun updateUserId(newValue: String)
+    external fun updateUserEmail(newValue: String)
+    external fun updateUserName(newValue: String)
+
+    /**
+     * Creates a new native bridge for interacting with native components.
+     * Configures logging and ensures that the reporting directory exists
+     * immediately.
+     */
+    init {
+        val outFile = File(reportDirectory)
+        if (!outFile.exists() && !outFile.mkdirs()) {
+            warn("The native reporting directory cannot be created.")
+        }
+    }
+
+    override fun update(observable: Observable, arg: Any?) {
+        if (isInvalidMessage(arg)) return
+
+        when (val msg = arg as StateEvent) {
+            is Install -> handleInstallMessage(msg)
+            DeliverPending -> deliverPendingReports()
```

Comment:
what is the difference between the entries in this pattern match that have `is` and those that don't?

Reply:
`is Install` checks whether `msg` is of type `Install`, whereas `DeliverPending` is a singleton so we only need to perform an equality check.